storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateEntry.java

```diff
@@ -41,18 +42,35 @@ public class ProducerStateEntry {
     private int coordinatorEpoch;
     private long lastTimestamp;
     private OptionalLong currentTxnFirstOffset;
+    
+    private VerificationState verificationState;
+    
+    // Before any batches are associated with the entry, the tentative sequence represents the lowest sequence seen.
+    private OptionalInt tentativeSequence;
+    
+    public enum VerificationState {
+        EMPTY,
+        VERIFYING,
+        VERIFIED
+    }
```

Comment:
Using a finite state machine for handling race conditions is susceptible to ABA problem, i.e. the logic may not distinguish between the original state not changing concurrently and changing back to the same state.  In this case it looks like the following sequence is possible:
1. Got message, started verifying transaction (transitioned to VERIFYING)
2. Transaction is in ongoing, so replied with success.
3. Transaction got aborted (transitioned to EMPTY).
4. Got another message, started verifying transaction (transitioned to VERIFYING)
5. The first message verification successfully completes, sees that state is VERIFYING and allows message to get written.

This is why we (and pretty much any distributed system) use ever-incrementing states (e.g. epochs) to handle concurrency, so that we don't get into ABA problem.

Reply:
this falls under the known gap for part 1. we can't prevent old batches from joining new transactions, but we can prevent hanging ones





[4:47](https://confluent.slack.com/archives/C055ADL2X9P/p1684280850696749)
once we get epoch bumps this will work correctly.

Reply:
As discussed offline, this race can cause hanging transactions:
1. producer properly adds partition to transaction
2. producer has trouble sending messages, retries a few times (all messages have the same epoch + sequence)
3. partition leader gets first message, sets state to VERIFYING, issues request to TC
4. TC sees the open transaction, replies with success
5. producer gives up and aborts transaction
6. partition leader gets abort marker, sets state to EMPTY
7. partition leader gets second message, sets state to VERIFYING, issues request to TC
8. successful reply from step 4 finally arrives, sees the state is VERIFYING, sets to VERIFIED
9. the first message is written on top of abort marker, resulting in "hanging" transaction

the issue is that transition VERIFYING - EMPTY - VERIFYING is the same as VERIFYING, the classic ABA problem https://en.wikipedia.org/wiki/ABA_problem

Reply:
I've addressed this issue by creating a verification object that is created on first attempt to verify and removed when a marker is written. 

When verification is needed, we pass this object through and check under log lock in the append path.
In the steps above, 6 will clear this object, 7 will set a new one, so 8 will not succeed and error out. 

When verification is not needed (already verified), we rely on firstTxnOffset being present before appending to the log.